src/runtime/TemporalObject.cpp

```diff
@@ -68,228 +68,268 @@ Value TemporalObject::toISOTime(ExecutionState& state, DateObject& d)
-        ErrorObject::throwBuiltinError(state, ErrorObject::RangeError, "Invalid time");
+    if (i != 0) {
+        index = i;
     }
 
-    return std::map<std::string, std::string>{ { "Year", year }, { "Month", month }, { "Day", day }, { "Hour", hour }, { "Minute", minute }, { "Second", second }, { "Millisecond", millisecond }, { "Microsecond", microSecond }, { "Nanosecond", nanoSecond }, { "Calendar", calendar } };
+    return timeZoneID;
 }
 
-std::map<std::string, std::string> TemporalObject::parseTemporalDateString(ExecutionState& state, std::string isoString)
+TemporalObject::DateTime TemporalObject::parseValidIso8601String(ExecutionState& state, std::string isoString)
 {
-    size_t pos = isoString.find('[');
-    std::string isoDate = isoString.substr(0, pos);
-    std::string extra;
+    DateTime dateTime = { 0, 0, 0, 0, 0, 0, 0, 0, 0, "" };
 
-    if (pos != std::string::npos) {
-        extra = isoString.substr(pos);
-    }
+    unsigned int index = 0;
 
-    pos = isoDate.find('T');
+    bool monthDay = false;
 
-    if (pos != std::string::npos) {
-        std::replace(isoDate.begin(), isoDate.end(), 'T', ' ');
-    }
+    try {
+        if (isoString[index] != 'T') {
+            // Date
+            if (isoString.rfind("−", index) == 0 && TemporalObject::is_number(isoString.substr(3 + index, 4))) {
+                dateTime.year = std::stoi("-" + isoString.substr(3 + index, 4));
+                index += 7;
+            } else if (TemporalObject::is_number(isoString.substr(index, 4))) {
+                dateTime.year = std::stoi(isoString.substr(0, 4));
+                index += 4;
+            } else if (isoString.rfind("--", 0) == 0) {
+                monthDay = true;
+                index += 2;
+            }
+
+            if (isoString[index] == '-') {
+                ++index;
+            }
 
-    pos = isoDate.find(' ');
-    std::string date = isoString.substr(0, pos);
-    std::string time = isoString.substr(pos + 1);
-    std::string year;
-    std::string month;
-    std::string week;
-    std::string day;
-    std::string timeZoneExtension;
-    std::string calendar;
-    pos = date.find('-');
-
-    if (pos != std::string::npos) {
-        if (pos == 0 && date.length() > 4) {
-            if (date.substr(pos + 1).find('-') == 0) {
-                date.erase(std::remove(date.begin(), date.end(), '-'), date.end());
-                if (date.length() == 4) {
-                    goto good;
+            if ((isoString[index] == '0' && std::isdigit(isoString[index + 1])) || (isoString[index] == '1' && isoString[index + 1] >= '0' && isoString[index + 1] <= '2')) {
+                dateTime.month = std::stoi(isoString.substr(index, 2));
+                index += 2;
+                if (isoString.empty()) {
+                    goto end;
```

Comment:
I think that `goto` operation makes the code complex and difficult to understand, so please replace these goto operations as much as possible